### PR TITLE
Group Membership Endpoint

### LINKED ISSUE #113: Group members endpoint
Looks like there's currently no way to handle group membership CRUD through the API.

My thought is that it probably makes sense to have this before a BP merge, at least if we plan to merge the other group-related endpoints. It feels to me like most use cases of the API (say, building a front-end interface) would need group membership functionality - at the very least, read-only to make the groups component useful.

### PR DESCRIPTION
**Pull request by renatonascalves**
fixes #113

### CODE REVIEW

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+			'description'       => __( 'Offset the result set by a specific number of pages of results.', 'buddypress' ),
+			'default'           => 1,
+			'type'              => 'integer',
+			'sanitize_callback' => 'absint',
+			'validate_callback' => 'rest_validate_request_arg',
+		);
+
+		$params['exclude'] = array(
+			'description'       => __( 'Ensure result set excludes specific member IDs.', 'buddypress' ),
+			'default'           => array(),
```
[the remark this answers is not in the document]
Comment (boonebgorges) [reply includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: A quick follow-up that I had to fix this before I could properly test. I added the following in `get_items()`:

```php

		if ( empty( $args['exclude'] ) ) {
			$args['exclude'] = false;
		}

```

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+			'type'              => 'array',
+			'sanitize_callback' => 'wp_parse_id_list',
+		);
+
+		$params['exclude_admins'] = array(
+			'description'       => __( 'Whether results should exclude group admins and mods.', 'buddypress' ),
+			'default'           => true,
```
Comment (boonebgorges) [inline includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: I wonder what you think about changing this default behavior. It's always confusing to me in BP that you don't get admins and mods by default. Third-party consumers will likely find it confusing as well. (Similar to #116, I think we should assume that clients have no prior knowledge of BP's PHP APIs, and we should take this opportunity to make the interface clearer.)

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+			'type'              => 'string',
+			'enum'              => array( 'last_joined', 'first_joined' ),
+			'sanitize_callback' => 'rest_validate_request_arg',
+		);
+
+		$params['roles'] = array(
+			'description'       => __( 'Ensure result set includes specific group roles.', 'buddypress' ),
+			'default'           => array(),
+			'type'              => 'array',
+			'sanitize_callback' => 'wp_parse_id_list',
```
Comment (boonebgorges) [inline includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: I believe that the roles are strings, not ints.

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+class BP_REST_Group_Members_Endpoint extends WP_REST_Controller {
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 0.1.0
+	 */
+	public function __construct() {
+		$this->namespace = 'buddypress/v1';
+		$this->rest_base = '/group/members';
```
Comment (boonebgorges) [inline includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: `/group/members` or `/groups/members/` ?

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+	/**
+	 * Check if a given request has access to group members.
+	 *
+	 * @since 0.1.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return bool
+	 */
+	public function get_items_permissions_check( $request ) {
+		return true;
```
Comment (boonebgorges) [inline includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: This should respect the privacy level of groups. Is there a way to inherit `can_see()` from the groups controller?

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+
+	/**
+	 * Check if a given request has access to update a group.
+	 *
+	 * @since 0.1.0
+	 *
+	 * @param  WP_REST_Request $request Full details about the request.
+	 * @return WP_Error|boolean
+	 */
+	public function update_item_permissions_check( $request ) {
```
Comment (boonebgorges) [inline includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: This will need to cover some more cases:

1. Groups admins can remove/ban/promote users
2. Users can remove themselves from groups
3. Users can add themselves to public groups
4. Users can request memberships to private groups (though I'm unsure whether this should be handled here or not)

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+	 *
+	 * @since 0.1.0
+	 *
+	 * @return array
+	 */
+	public function get_update_collection_params() {
+		$params                       = parent::get_collection_params();
+		$params['context']['default'] = 'view';
+
+		$params['id'] = array(
```
Comment (boonebgorges) [inline includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: IMO this should be `user_id`. Group memberships have unique IDs that are distinct from the user IDs, and calling this param `id` could cause confusion. It also is somewhat unclear given that we need to pass `group_id`.

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+		if ( (int) $id <= 0 ) {
+			return false;
+		}
+
+		$user = get_userdata( (int) $id );
+		if ( empty( $user ) || ! $user->exists() ) {
+			return false;
+		}
+
+		if ( is_multisite() && ! is_user_member_of_blog( $user->ID ) ) {
```
Comment (boonebgorges) [inline includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: I think that, on many multisite setups, BP users do not have a role on the main site.

File: tests/groups/test-group-member-controller.php [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+
+		$request->set_param( 'context', 'view' );
+		$response = $this->server->dispatch( $request );
+
+		$this->assertEquals( 200, $response->get_status() );
+
+		$all_data = $response->get_data();
+		$data     = $all_data;
+
+		foreach ( $all_data as $data ) {
```
Comment (boonebgorges) [inline tests/groups/test-group-member-controller.php]: These tests are all passing locally for me even though they're returning no members (see my comment about `exclude` above) and I think it's because the `foreach` loop isn't triggered when `$all_data` is empty. Putting `assertNotEmpty()` before is probably sufficient.

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php [hunk 10]
```diff
@@ -0,0 +1,517 @@
...
+			'sanitize_callback' => 'absint',
+			'validate_callback' => 'rest_validate_request_arg',
+		);
+
+		$params['status'] = array(
+			'description'       => __( 'Sort the order of results by the status of the group members.', 'buddypress' ),
+			'default'           => 'last_joined',
+			'type'              => 'string',
+			'enum'              => array( 'last_joined', 'first_joined' ),
+			'sanitize_callback' => 'rest_validate_request_arg',
```
Comment (boonebgorges) [inline includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: This should be `validate_callback`. This typo is causing tests to fail, because the `sanitize_callback` is causing a chain-reaction that's triggering a `type=online` query, which always returns no members during unit testing :-D

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php (hunk 4 above)
Comment (renatonascalves) [reply includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: @boonebgorges I think `/group/members`, singular, as we are fetching members of a specific, single, group instead of members of several groups.

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php (hunk 8 above)
Comment (renatonascalves) [reply includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: No caps, no nothing? `is_user_member_of_blog` doesn't check roles, but caps.

Comment (boonebgorges) [reply includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: Correct. It looks like `bp_members_add_role_after_activation()` was added in 3.0.0 to fix this, but prior to this (and so affecting many older installations) it's possible to be a BP member without being a main-site user. See https://buddypress.trac.wordpress.org/ticket/7565.

I'd lean toward removing this check.

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php (hunk 4 above)
Comment (boonebgorges) [reply includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: Gotcha. Neither one seems great to me, to be honest :) When read in a RESTful way, `group/members/` implies that there's a group called `members`. What do you think of removing the (pseudo-)nesting - `$rest_base = '/group-members'` ?

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (boonebgorges) [reply includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: I've opened a BP ticket describing the issue: https://buddypress.trac.wordpress.org/ticket/8040#ticket Up to you whether you want to change the behavior here too.

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php [hunk 11]
```diff
@@ -0,0 +1,517 @@
...
+
+	/**
+	 * Get the group member schema, conforming to JSON Schema.
+	 *
+	 * @since 0.1.0
+	 *
+	 * @return array
+	 */
+	public function get_item_schema() {
+		$members_endpoint = new BP_REST_Members_Endpoint();
```
[the remark this answers is not in the document]
Comment (renatonascalves) [reply includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: Added! :)

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php (hunk 8 above)
Comment (renatonascalves) [reply includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: Got it! Removing it! :)

File: includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php (hunk 6 above)
Comment (renatonascalves) [reply includes/bp-groups/classes/class-bp-rest-group-members-endpoint.php]: @boonebgorges 1, 2 and 3 were added. Number 4 will be tackled later!

**Comment by boonebgorges** [comment]
Thanks, @renatonascalves ! Your logic was mostly right. I added a clause ensuring that users are not able to add other users to public groups.

I found it quite hard to follow the permission checks the way you'd written them, so I had a go at reorganizing. I put the `bp_moderate` check first. Then, I handled self-requests (user_id === loggedin user) separately from requests about other users. I think this simplifies the flow.

Note that, in the `default` case, I'm just returning `false`. This throws the regular `rest_forbidden` error instead of a custom one. I don't feel strongly about this - it makes the logic much simpler not to have a separate error for each failure, but it also provides less info for the client. Also, I noticed you're serving 500s for these custom errors, but the core endpoints use `rest_authorization_required_code()`, which is `403`. Is there a reason for this?
